/*********************************************************************************
 * Modifications Copyright 2017-2019 eBay Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 *
 *********************************************************************************/

syntax = "proto3";

package nuraft_mesg;

enum LogType {
    UNKNOWN = 0;
    APP = 1;
    CONFIG = 2;
    CLUSTER = 3;
    PACK = 4;
    SNAPSYNC = 5;
    CUSTOM = 999;
}

enum ResultCode {
    OK = 0;
    CANCELLED = 1;
    TIMEOUT = 2;
    NOT_LEADER = 3;
    BAD_REQUEST = 4;
    SERVER_ALREADY_EXISTS = 5;
    CONFIG_CHANGING = 6;
    SERVER_IS_JOINING = 7;
    SERVER_NOT_FOUND = 8;
    CANNOT_REMOVE_LEADER = 9;
    SERVER_IS_LEAVING = 10;
    TERM_MISMATCH = 11;

    RESULT_NOT_EXIST_YET = 10000;

    FAILED = 32768;
}

message LogEntry {
    uint64 term = 1;
    LogType type = 2;
    bytes buffer = 3;
    uint64 timestamp = 4;
}

message RCMsgBase {
    uint64 term = 1;
    uint32 type = 2;
    int32 src = 3;
    int32 dest = 4;
}

message RCRequest {
    uint64 last_log_term = 1;
    uint64 last_log_index = 2;
    uint64 commit_index = 3;
    repeated LogEntry log_entries = 4;
}

message RCResponse {
    uint64 next_index = 1;
    bool accepted = 2;
    bytes context = 3;
    string dest_addr = 4;
    ResultCode result_code = 5;
}

message RaftMessage {
    RCMsgBase base = 1;
    oneof req_resp {
        RCRequest rc_request = 2;
        RCResponse rc_response = 3;
    }
}
